[dockerfile] Allow ARG in FROM

dnephin opened the pull request:
Fixes #18119

This has been a major pain point for building from common base images. There's no way to build a base image with a unique tag (git sha is a common one), and then build app specific images from that tag.

dnephin committed:
Allow ARG to come before FROM to support variables in FROM.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

duglin commented:
ping @jhowardmsft @tianon

duglin commented:
Definitely +1 to the idea, but we'll need lots of tests :-)

dnephin commented on builder/dockerfile/dispatchers.go:
```diff
@@ -769,6 +780,10 @@ func arg(b *Builder, args []string, attributes map[string]bool, original string)
+	if b.image == "" {
+		return nil
+	}
 	return b.commit("", b.runConfig.Cmd, fmt.Sprintf("ARG %s", arg))
```
I see that something like this was discussed already in #29766. I think we could go with either approach.

griwes commented:
As the author of #18119, +100.

tonistiigi commented:
Design LGTM

thaJeztah commented:
Haven't checked exactly what's possible with this PR yet, but adding some points that came up in previous discussions;

- Dockerfile should always be buildable without specifying `--build-arg`, this probably requires that the `ARG` _must_ have a default. Empty values _could_ be an option, but what happens with `FROM foo:$tag`, which leads to `FROM foo:` (trailing colon)?
- Do we allow both _name_ / registry, and _tag_ to be set? `FROM $registry:$port/$repo:$tag`, or just _tag_?
- Having a dedicated option for this, instead of `ARG` came up at some point. This _would_ give us more flexibility, and control over what's possible / acceptable, but TBD if that's needed/desirable

thaJeztah commented:
With docker 1.13+ doing API version negotiation, we may have to throw an error if this feature is tried against an older daemon, but that's orthogonal to overall design

duglin commented:
I think it will anyway since it'll flag a pre-FROM cmd

StefanScherer commented:
That is good for creating images for different architectures. So I don't need two Dockerfiles, one with FROM node:7.4.0-slim and another with hypriot/rpi-node:7.4.0-slim or an Windows image if the rest is identical.

friism commented:
cc @tt https://github.com/docker/docker/issues/18585

justincormack commented:
@thaJeztah your points only seem to work if you have a `docker build --from alpine:3.4 .` option say, which completely overrides the specified `FROM`. That seems a little unintuitive, as the Dockerfile is specifying one thing but the build does another, but maybe it is ok, it is not really much different from `ARG`.

cpuguy83 commented:
Yeah, originally I was thinking a specific `--from` flag personally... but having it parameterized with what we have now is not awful.

tonistiigi commented:
@thaJeztah Can we move this to code review?

tonistiigi commented on builder/dockerfile/dispatchers.go:
```diff
@@ -769,6 +780,10 @@ func arg(b *Builder, args []string, attributes map[string]bool, original string)
 	}
 
+	// Arg before FROM doesn't add layers
+	if b.image == "" {
```
`&& !b.noBaseImage`

thaJeztah commented:
Yes, let's move if forward

tonistiigi commented:
Why does this use a different variable replacer than other commands?

duglin commented:
can we get some tests added?

AkihiroSuda commented:
Warning seems weird
```
suda@ws01:/tmp/a> cat Dockerfile 
FROM busybox:$foo

suda@ws01:/tmp/a> docker build --build-arg foo=latest .
Sending build context to Docker daemon  2.048kB
Step 1/1 : FROM busybox:$foo
 ---> 7968321274dc
[Warning] One or more build-args [foo] were not consumed
Successfully built 7968321274dc
```

duglin commented:
Using:
```
FROM ubuntu$version
RUN echo hi
```
and `docker build --build-arg version=:foo .`
I get:
```
Sending build context to Docker daemon  17.92kB
Step 1/2 : FROM ubuntu$version
manifest for ubuntu:foo not found
```
Which would be correct if I had `ARG version` before the FROM.  Even though I've never liked that we had to define build args via an ARG, since that's the rule, we should instead seen this error:
`[Warning] One or more build-args [version] were not consumed`.

duglin commented:
nvm, my mistake

duglin commented:
@AkihiroSuda the error you're seeing is correct since there is no ARG defined for "version"

dnephin commented:
> Why does this use a different variable replacer than other commands?

> can we get some tests added?

Yup, I was just waiting for this to get into code-review before investing time in tests (mentioned as a TODO in the description). I'll get to adding those now.

> Warning seems weird

I'll look into it

AkihiroSuda commented:
Thank you for replying, but I think the PR has still some weird behavior.
Even when there is no `ARG foo`, `$foo` seems somehow consumed.

```
$ cat Dockerfile 
# since there is no `ARG foo`, $foo should never be used.
FROM busybox:$foo
```

```
$ docker build --build-arg foo=nonexistent .
Sending build context to Docker daemon  2.048kB
Step 1/1 : FROM busybox:$foo
manifest for busybox:nonexistent not found
```

alexellis commented:
From recollection, doesn't `--build-arg` for the time-being only cover proxy-related args? If that's still the case there is some new/pending by @dave-tucker around that area.

dnephin commented:
The only "default" build-args are related to proxies, but the user can define arbitrary build-args.

dhiltgen commented:
I think this would make a nice compliment to #32063 and enable quite a few projects to clean up messy scripts and clunky external rigging to build nice clean compact final production images.

It'd be slick if we could get this merged in time to make the same release as the new multi-stage build support.

dnephin committed:
Fix arg in from when arg is not defined

Add mock builder backend
Add tests for ARG in FROM

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin commented:
Ok, this is rebased, and I think I've fixed the issues.

@AkihiroSuda I fixed the "using arg without it being defined", and added a test

@tonistiigi I reverted my changes to shellvar, and it now uses `ProcessWord()` like everything else

I've added tests, along with a `MockBackend`

dnephin commented:
windowsRS1 failure should be unrelated

vieux commented:
LGTM

cpuguy83 commented on builder/dockerfile/dispatchers_test.go:
```diff
@@ -189,35 +191,68 @@ func TestLabel(t *testing.T) {
-			t.Fatalf("Error when executing from: %s", err.Error())
-		}
+func TestFromWithArg(t *testing.T) {
+	tag, expected := ":latest", "expectedthisid"
```
Can tag be something other than latest?

dnephin replied:
ok

cpuguy83 commented on builder/dockerfile/dispatchers_test.go:
```diff
@@ -189,35 +191,68 @@ func TestLabel(t *testing.T) {
+}
+
+func TestFromWithUndefinedArg(t *testing.T) {
+	tag, expected := "latest", "expectedthisid"
```
Same, something other than latest.

cpuguy83 commented on builder/dockerfile/mockbackend_test.go:
```diff
@@ -0,0 +1,102 @@
+	return i.config
+}
+
+var _ builder.Image = &mockImage{}
```
What are these for?

dnephin replied:
This is to assert and document that the mock implements the interface. I can remove it if it's an issue.

cpuguy83 commented on builder/dockerfile/shell_parser_test.go:
```diff
@@ -54,7 +54,7 @@ func TestShellParser4EnvVars(t *testing.T) {
 			} else {
 				assert.NilError(t, err)
-				assert.DeepEqual(t, newWord, []string{expected})
+				assert.DeepEqual(t, newWord, expected)
```
If this is just a string, can we do `assert.Equal` instead?

dnephin replied:
yup, can do that now

tonistiigi commented:
This does not work for the multi-stage build case. Args can't be used in subsequent `FROM` commands and defaults before first `FROM` don't carry to the `ARG` after that. https://github.com/docker/docker/commit/d3b6b77875270f3db031f82e1250b5078d613936 is a commit from the time that multi-stage build was on top of this PR and had support for that. That commit also includes a test for this case.

This also needs some actual integration tests to prove the feature actually works and doesn't regress. I'm really not a fan of these tests that set up some private structure, then call a private method and test if a call to that method changed a private property inside that struct. They don't test any feature, only some hidden implementation detail.

dnephin commented:
> Args can't be used in subsequent FROM commands and defaults before first FROM don't carry to the ARG after that... That commit also includes a test for this case.

Thanks, I'll try it out.

> I'm really not a fan of these tests that set up some private structure, then call a private method and test if a call to that method changed a private property inside that struct. They don't test any feature, only some hidden implementation detail.

These functions aren't private... they're just not exported. The surface area of this package is massive. There's no way you can write unit tests by only testing the exported functions. Attempting to test all the behaviour with integration tests is brutally slow.

I would be nice to be able to only test the exported functions, but unfortunately the code isn't designed to support that kind of testing. 

I agree that having an integration test for the high level features is important, but that doesn't remove the need for faster unit tests.

tonistiigi commented:
@dnephin I totally agree that `builder` pkg or its executor dependency is not well designed for unit testing. It's not only exported/unexported. The assertion for `ARG` is that correct values and data from correct `FROM` image end up in exported image metadata and `RUN` commands. Not that something ends up in `b.allowedBuildArgs` map. Atm, only way this can be asserted is in integration tests.

dnephin committed:
Refactor BuildArgs

Add MetaArgs for ARG that occur before the first FROM
Integration test for these cases.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin commented:
@cpuguy83  addressed your comments

@tonistiigi added your integration test with the changes we discussed (`$tag` on second `FROM`)

Last commit extracts all the build arg related logic into a new struct, and adds handling for the "arg before from" special case. I've called these `MetaArg` for now, but I'd be happy to change it for a better name.

tonistiigi commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -4745,29 +4745,45 @@ func (s *DockerSuite) TestBuildBuildTimeArgMultipleFrom(c *check.C) {
+	c.Assert(result.Stdout(), checker.Not(checker.Contains), "tag")
+
+	result = cli.DockerCmd(c, "run", "--rm", imgName, "cat", "/out")
+	c.Assert(result.Stdout(), checker.Contains, "tag=latest")
```
Could you add a testcase that tags busybox to `nosuctag` and  doesn't set `--build-arg` to confirm that the defaults apply correctly.

dnephin replied:
I can, although I think that behaviour is already covered by unit tests.

tonistiigi commented:
Small comment about the test, otherwise LGTM

vdemeester reviewed (approved):
LGTM 🦁

cpuguy83 reviewed (approved):
LGTM

cpuguy83 commented:
Moved to docs review.

thaJeztah commented:
@dnephin can you update the `builder` documentation?

- This section mentions that `FROM` must be the _first_ command in a Dockerfile; https://github.com/docker/docker/blob/master/docs/reference/builder.md#format
- Possibly there needs to be a mention in the `ARG` section, explaining that it can be used in `FROM` (and that it's scoped per `FROM` IIRC) https://github.com/docker/docker/blob/master/docs/reference/builder.md#arg
- Likely the `FROM` section; https://github.com/docker/docker/blob/master/docs/reference/builder.md#from

Let me know if you want to do that as part of this PR, or as a follow-up

dnephin commented:
Thanks! I will do the docs in a follow up PR

thaJeztah commented:
Ok, let's go ahead, merge this, and do the docs during code-freeze 👍

hillct commented:
@dnephin @thaJeztah While the documentation was updated https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact it fails every time, using the documented format (in my case, for Alpine version as seen here: https://hub.docker.com/r/pinnacledigital/docker-alpine-abuild/~/dockerfile/ ) Is this something that didn't make it from Docker-Engine to the new dockerd architecture?

Note, it also fails locally, as tested using the latest release for Mac, with, or without use of the defined build hook.

dnephin commented:
It's in 17.05 and 17.06

thaJeztah commented:
@hillct if you want to use it for automated builds; Docker Hub, nor Docker Cloud are updated yet to 17.06 (they generally follow the EE releases); in Docker Cloud, you can configure a custom engine version though

hillct commented:
@dnephin I checked my version and find I'm running
Version 17.06.0-ce-mac18 (18433)
Channel: stable
d9b66511e0

The error I receive is as follows:

Sending build context to Docker daemon   89.6kB
Step 1/16 : ARG linux_versiom=latest
 ---> 
Step 2/16 : FROM alpine:${linux_version}
invalid reference format

Foxlik commented:
@hillct there is a typo in your first step ... replace m with n and try again ;-)

hillct commented:
@fortinux It's always the silliest things. Earlier iterations of this Dockerfile included ARG VERSION=latest and ARG LINUX_VERSION=latest both of which failed, which seems to possibly relate to use of capital letters in the variable name, in FROM even if the value of the variable is in lower case. Not sure if this is an error/holdover around FROM directive validation, or intentional, and merely requiring clarification in documentation.

rogaha commented:
thanks a lot for adding this @dnephin! It's awesome!

olkgrab commented:
Hello,

Looks like don't work for me. Here is my Dockerfile:
`FROM node:4.4.7
ARG VERSION_DAS=4.4.7
RUN echo $VERSION_DAS`

run success 
`Sending build context to Docker daemon  68.57MB
Step 1/3 : FROM node:4.4.7
 ---> 93b396996a16
Step 2/3 : ARG VERSION_DAS=4.4.7
 ---> Running in a088e6265995
 ---> a916acc1199a
Removing intermediate container a088e6265995
Step 3/3 : RUN echo $VERSION_DAS
 ---> Running in 487025b76de9
4.4.7
 ---> a98702da95b7
Removing intermediate container 487025b76de9
Successfully built a98702da95b7`

But when I move ARG before FROM, it is didn't work
`ARG VERSION_DAS=4.4.7
FROM node:4.4.7
RUN echo $VERSION_DAS
`
Result
`docker build -f Dockerfile .
Sending build context to Docker daemon  68.57MB
Step 1/3 : ARG VERSION_DAS=4.4.7
 ---> 
Step 2/3 : FROM node:4.4.7
 ---> 93b396996a16
Step 3/3 : RUN echo $VERSION_DAS
 ---> Using cache
 ---> 2b731d6f09c2
Successfully built 2b731d6f09c2`

Docker version
`Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:20:36 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:21:56 2017
 OS/Arch:      linux/amd64
 Experimental: false`

Could you please advice, am I doing something wrong?

tonistiigi commented:
@olkgrab Every build stage has to define its own args. If no default has been set for a stage it inherits it from the "global FROM" args.

```
ARG VERSION_DAS=4.4.7
FROM node:4.4.7
ARG VERSION_DAS
RUN echo $VERSION_DAS
```

rogaha commented:
@tonistiigi this is the second time I see people misunderstanding this behavior (e.g. https://github.com/moby/moby/issues/34008), we need to improve the docs to make these details are well explained.

dnephin commented:
This seems pretty clear to me: https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact

> FROM instructions support variables that are declared by any ARG instructions that occur before the first FROM.

dnephin commented:
How's this look https://github.com/docker/cli/pull/333 ?

thehesiod commented:
it would be nice if there was a warning, the behavior was really strange w/o any warnings

arminanton commented:
Would it be possible to allow variables (ARG ones) to be used after `AS`, like in `FROM base AS ${service_name}`?

thaJeztah commented:
- @arminanton see https://github.com/moby/buildkit/issues/2618
